File: internal/plugins/jest/rules/no_restricted_jest_methods/no_restricted_jest_methods_test.go
package no_restricted_jest_methods_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_restricted_jest_methods"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoRestrictedJestMethodsRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_restricted_jest_methods.NoRestrictedJestMethodsRule,
		[]rule_tester.ValidTestCase{
			{Code: `jest`},
			{Code: `jest()`},
			{Code: `jest.mock()`},
			{Code: `expect(a).rejects;`},
			{Code: `expect(a);`},
			{
				Code: `import { jest } from '@jest/globals';

jest;
`,
			},
			{
				Code: `const jest = { fn: () => {} };
jest.fn();`,
				Options: []interface{}{
					map[string]interface{}{"fn": nil},
				},
			},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `jest.fn()`,
				Options: []interface{}{
					map[string]interface{}{"fn": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `fn` is disallowed",
						Line:      1,
						Column:    6,
					},
				},
			},
			{
				Code: `jest["fn"]()`,
				Options: []interface{}{
					map[string]interface{}{"fn": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `fn` is disallowed",
						Line:      1,
						Column:    6,
					},
				},
			},
			{
				Code: `jest.fn()`,
				Options: []interface{}{
					map[string]interface{}{"fn": ""},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `fn` is disallowed",
						Line:      1,
						Column:    6,
					},
				},
			},
			{
				Code: `jest.fn().mockImplementation(() => {})`,
				Options: []interface{}{
					map[string]interface{}{"fn": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `fn` is disallowed",
						Line:      1,
						Column:    6,
					},
				},
			},
			{
				Code: `(jest.fn)()`,
				Options: []interface{}{
					map[string]interface{}{"fn": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `fn` is disallowed",
						Line:      1,
						Column:    7,
					},
				},
			},
			{
				Code: `jest.mock()`,
				Options: []interface{}{
					map[string]interface{}{"mock": "Do not use mocks"},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethodWithMessage",
						Message:   "Do not use mocks",
						Line:      1,
						Column:    6,
					},
				},
			},
			{
				Code: `jest["mock"]()`,
				Options: []interface{}{
					map[string]interface{}{"mock": "Do not use mocks"},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethodWithMessage",
						Message:   "Do not use mocks",
						Line:      1,
						Column:    6,
					},
				},
			},
			{
				Code: `import { jest } from '@jest/globals';

jest.advanceTimersByTime();
`,
				Options: []interface{}{
					map[string]interface{}{"advanceTimersByTime": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `advanceTimersByTime` is disallowed",
						Line:      3,
						Column:    6,
					},
				},
			},
			{
				Code: `const { jest } = require('@jest/globals');

jest.advanceTimersByTime();
`,
				Options: []interface{}{
					map[string]interface{}{"advanceTimersByTime": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedJestMethod",
						Message:   "Use of `advanceTimersByTime` is disallowed",
						Line:      3,
						Column:    6,
					},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-restricted-jest-methods.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-restricted-jest-methods', {} as never, {
  valid: [
    { code: 'jest' },
    { code: 'jest()' },
    { code: 'jest.mock()' },
    { code: 'expect(a).rejects;' },
    { code: 'expect(a);' },
    {
      code: `
        import { jest } from '@jest/globals';

        jest;
      `,
    },
  ],
  invalid: [
    {
      code: 'jest.fn()',
      options: [{ fn: null }],
      errors: [
        {
          messageId: 'restrictedJestMethod',
          data: {
            message: null,
            restriction: 'fn',
          },
          column: 6,
          line: 1,
        },
      ],
    },
    {
      code: 'jest["fn"]()',
      options: [{ fn: null }],
      errors: [
        {
          messageId: 'restrictedJestMethod',
          data: {
            message: null,
            restriction: 'fn',
          },
          column: 6,
          line: 1,
        },
      ],
    },
    {
      code: 'jest.mock()',
      options: [{ mock: 'Do not use mocks' }],
      errors: [
        {
          messageId: 'restrictedJestMethodWithMessage',
          data: {
            message: 'Do not use mocks',
            restriction: 'mock',
          },
          column: 6,
          line: 1,
        },
      ],
    },
    {
      code: 'jest["mock"]()',
      options: [{ mock: 'Do not use mocks' }],
      errors: [
        {
          messageId: 'restrictedJestMethodWithMessage',
          data: {
            message: 'Do not use mocks',
            restriction: 'mock',
          },
          column: 6,
          line: 1,
        },
      ],
    },
    {
      code: `
        import { jest } from '@jest/globals';

        jest.advanceTimersByTime();
      `,
      options: [{ advanceTimersByTime: null }],
      errors: [
        {
          messageId: 'restrictedJestMethod',
          data: {
            message: null,
            restriction: 'advanceTimersByTime',
          },
          column: 6,
          line: 3,
        },
      ],
    },
  ],
});
